pkg/controllers/cluster/actions/rack_synchronized.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (C) 2021 ScyllaDB
+
+package actions
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"github.com/scylladb/go-log"
+	scyllav1 "github.com/scylladb/scylla-operator/pkg/api/v1"
+	"github.com/scylladb/scylla-operator/pkg/controllers/cluster/util"
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+)
+
+type rackSynchronizedSubAction interface {
+	// RackUpdated should return whether requested update is already applied to a rack sts.
+	RackUpdated(sts *appsv1.StatefulSet) (bool, error)
+	// Update performs desired update.
+	Update(sts *appsv1.StatefulSet) error
+	// Name of action.
+	Name() string
+}
+
+// rackSynchronizedAction can be used to perform an multi rack update in synchronous manner.
+// Racks are upgraded in the same order as in ScyllaCluster Spec.
+// Single Execute call performs at most a single update.
+type rackSynchronizedAction struct {
+	subAction rackSynchronizedSubAction
+	cluster   *scyllav1.ScyllaCluster
+	logger    log.Logger
+}
+
+const (
+	RackSynchronizedActionPrefix = "rack-synchronized-"
+)
+
+func (a rackSynchronizedAction) Name() string {
+	return fmt.Sprintf("%s%s", RackSynchronizedActionPrefix, a.subAction.Name())
+}
+
+func (a rackSynchronizedAction) Execute(ctx context.Context, s *State) error {
+	for _, rack := range a.cluster.Spec.Datacenter.Racks {
+		sts := &appsv1.StatefulSet{}
+		sts, err := util.GetStatefulSetForRack(ctx, rack, a.cluster, s.kubeclient)
+		if err != nil {
+			return errors.Wrap(err, "get rack statefulset")
+		}
+
+		rackUpdated, err := a.subAction.RackUpdated(sts)
+		if err != nil {
+			return errors.Wrap(err, "determine if rack needs update")
+		}
+		if !rackUpdated {
+			if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
this is equivalent to the patch. you avoid checking stuff like `deletionTimestamp!=nil`, ownerRefs, UID, ... after the change. If you hit conflict, exit the sync loop and when informer sees the new version it will run a new sync loop